pkg/pool-stable/contracts/ComposableStablePool.sol

```diff
@@ -934,10 +934,26 @@ contract ComposableStablePool is
         _upscaleArray(balancesIncludingBpt, _scalingFactors());
 
         (uint256 virtualSupply, uint256[] memory balances) = _dropBptItemFromBalances(balancesIncludingBpt);
+        (uint256 lastJoinExitAmp, uint256 lastPostJoinExitInvariant) = getLastJoinExitData();
+        (uint256 protocolOwnershipPercentage, uint256 oldAmpInvariant) = _getProtocolPoolOwnershipPercentage(
+            balances,
+            lastJoinExitAmp,
+            lastPostJoinExitInvariant
+        );
 
+        // We want to multiply the rate by the fraction of the pool which is not owed to the protocol on the next
+        // collection of protocol fees to reduce the value of BPT accordingly, otherwise the value of BPT can be
+        // manipulated by timing collections of protocol fees.
         (uint256 currentAmp, ) = _getAmplificationParameter();
-
-        return StableMath._getRate(balances, currentAmp, virtualSupply);
+        if (currentAmp == lastJoinExitAmp) {
+            // If the amplitude hasn't changed then we can reuse the already calculated invariant.
+            return oldAmpInvariant.divDown(virtualSupply).mulDown(protocolOwnershipPercentage.complement());
+        } else {
+            return
+                StableMath._getRate(balances, currentAmp, virtualSupply).mulDown(
+                    protocolOwnershipPercentage.complement()
+                );
```

Comment:
Instead of using `complement` and reasoning about it this way, I'd simply call

```
uint256 protocolFeeAmount = _calculateAdjustedProtocolFeeAmount(
    virtualSupply,
    expectedProtocolOwnershipPercentage
);
```

and add that to `virtualSupply`, to get the _real_ virtual supply. Those BPT should be minted - the only reason why it hasn't happened yet is due to gas optimizations favoring regular swaps. They will in fact be minted before the next interaction that involves BPT (and hence update the virtual supply).

Reply:
So this could become

```
uint256 currentInvariant = (currentAmp == lastJoinExitAmp) ? 
  oldAmplnvariant : 
  StableMath.calculateInvariant(currentAmp, balances);

return currentInvariant.div(virtualSupply.add(protocolFeeAmount));
```

(`StableMath.getRate` is just that division)

Reply:
Note that this clearly shows that updating the protocol fee rate cache manipulates the rate - we need to decide if that's an issue.

Reply:
Actually, I'd consider changing `getVirtualSupply` as well to include this